### PR TITLE
ui/cdc: Highwater timestamp on Jobs page

### PR DESCRIPTION
**Pull request by mrtracy**
Adds support to the jobs page for jobs which have a "highwater
timestamp" instead of a "fraction completed".

+ Add highwater timestamp fields to the Admin API Jobs method. There are
actually two fields: one returns the value as a timestamp, the other
returns a string which is the server-side decimal representation of the
time, appropriate for use in "AS OF SYSTEM TIME" statements.
+ Jobs page now displays "Last Processed" time in the Status column for
jobs which have a highwater timestamp instead of a fraction completed.
The time string is displayed using moment.js's "fromNow()" style.
+ Added a tooltip when a user hovers over a "Last Processed" which
displays the decimal-format system time.

Release note (admin ui change): The Jobs page now supports
indefinitely-running job types which have a "highwater timestamp",
instead of the "fraction completed" used by jobs with a finite task.

screen shot 2018-08-13 at 11 52 54 am

### CODE REVIEW

**Review by danhhz — COMMENTED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28535#-:-LJorIGO2Y1P7pcWOaa4:bg7igkv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/server/serverpb/admin.proto, line 456 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28535#-LJoqrDQ0xjcGn8B5R40:-LJoqrDQ0xjcGn8B5R41:b-qmhks0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1505177824f8ca752c9ea110f6ee0898471fc034/pkg/server/serverpb/admin.proto#L456)):*
> ```protobuf
>     float fraction_completed = 11;
>     string error = 12;
>     // highwater_timestamp is the highwater timestamp returned as an HLC
> ```

nit: not an HLC right?
___
*[pkg/ui/src/views/jobs/index.tsx, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28535#-LJor0TkA6lDEBlEuX0k:-LJor0TkA6lDEBlEuX0l:biboba) ([raw file](https://github.com/cockroachdb/cockroach/blob/1505177824f8ca752c9ea110f6ee0898471fc034/pkg/ui/src/views/jobs/index.tsx#L53)):*
> ```XML
>   { value: jobType.IMPORT.toString(), label: "Imports" },
>   { value: jobType.SCHEMA_CHANGE.toString(), label: "Schema Changes" },
>   { value: jobType.CHANGEFEED.toString(), label: "Change Feed"},
> ```

nit: we've been styling this as one word "changefeed" in the docs: https://www.cockroachlabs.com/docs/v2.1/create-changefeed.html
___
*[pkg/ui/src/views/jobs/index.tsx, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28535#-LJorMEh3UTmEDne7jJf:-LJorMEh3UTmEDne7jJg:bi72bjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1505177824f8ca752c9ea110f6ee0898471fc034/pkg/ui/src/views/jobs/index.tsx#L130)):*
> ```XML
>     return (
>       <ToolTipWrapper text={`System Time: ${tooltip}`}>
>         Last Processed: {highwaterMoment.fromNow()}
> ```

"Last Processed" seems fuzzy to me, but if you don't like high-water, I guess I'm okay with it

File: pkg/ui/src/views/jobs/index.tsx [hunk 1]
```diff
@@ -18,8 +18,10 @@ import { PageConfig, PageConfigItem } from "src/views/shared/components/pageconf
 import { SortSetting } from "src/views/shared/components/sortabletable";
 import { ColumnDescriptor, SortedTable } from "src/views/shared/components/sortedtable";
 import { ToolTipWrapper } from "src/views/shared/components/toolTip";
+import { NanoToMilli } from "src/util/convert";
 
 import spinner from "assets/spinner.gif";
+import { converter } from "../../../node_modules/protobufjs";
```
Comment (vilterp) [inline pkg/ui/src/views/jobs/index.tsx]: Can you just `import … from "protobufjs"`? Weird to see `node_modules` here.

**Review by vilterp — COMMENTED** [review]
Q: how often would we expect this to update on the backend (i.e. how frequently does the high water timestamp change) vs how often does this jobs page autorefresh? I suppose it will pretty much always say "a few seconds ago" for a running changefeed?

**Review by mrtracy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28535)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/server/serverpb/admin.proto, line 456 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28535#-LJoqrDQ0xjcGn8B5R40:-LK-9tb_CBMTS423LF_Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1505177824f8ca752c9ea110f6ee0898471fc034/pkg/server/serverpb/admin.proto#L456)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: not an HLC right?
</blockquote></details>

Done.
___
*[pkg/ui/src/views/jobs/index.tsx, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28535#-LJoeWyBWLZDSjq0k27--r1-24:-LK-AB3E7kp7lzhwPLxO:b-frd2j1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1505177824f8ca752c9ea110f6ee0898471fc034/pkg/ui/src/views/jobs/index.tsx#L24)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

Can you just `import … from "protobufjs"`? Weird to see `node_modules` here.
</blockquote></details>

That must have been automatically added? No idea, removed.
___
*[pkg/ui/src/views/jobs/index.tsx, line 53 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28535#-LJor0TkA6lDEBlEuX0k:-LK-AAUO2sFofdgc9IIO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1505177824f8ca752c9ea110f6ee0898471fc034/pkg/ui/src/views/jobs/index.tsx#L53)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: we've been styling this as one word "changefeed" in the docs: https://www.cockroachlabs.com/docs/v2.1/create-changefeed.html
</blockquote></details>

Done.
___
*[pkg/ui/src/views/jobs/index.tsx, line 130 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28535#-LJorMEh3UTmEDne7jJf:-LK-AJH_EsiceIXXzAbn:btdjirc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1505177824f8ca752c9ea110f6ee0898471fc034/pkg/ui/src/views/jobs/index.tsx#L130)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

"Last Processed" seems fuzzy to me, but if you don't like high-water, I guess I'm okay with it
</blockquote></details>

Changed to "High-water timestamp:"

**Review by mrtracy — COMMENTED** [review]
I suppose it probably will always say "a few seconds ago" except when there is a problem.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28535#-:-LK-CAp1Djn_T_dbzVwO:bb8nran)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Review by vilterp — APPROVED** [review]

**Comment by couchand** [comment]
There are a few things unfortunate here:

- there's a conceptual difference between changefeeds and all other jobs.  this is most evident in the description column: it's listed as `CREATE CHANGEFEED ...`, which next to a schema change or import/export job makes it seem like the act of creating the changefeed takes time.  I think it's important to do more to make this clear.

- labeling it "High-water timestamp" (which I agree with because of the principle of Ubiquitous Language) makes rendering it as a relative time a bit inconsistent.

- manually clamping it to the client's clock just seems wrong (not using a relative time solves this reasonably)

- if the point of the adding the decimal form is to enable users to use it in an `AS OF SYSTEM TIME` query, putting it in a tooltip defeats the purpose, since it can't be copy-pasted easily.

This really would have been nice to have the design discussed ahead of implementation.

**Pull request merged**